test/tester.go

```diff
@@ -0,0 +1,206 @@
+	// instance of the service and provides simplified ways to call the handlers.
+	AccountsTester struct {
+		Ctx    context.Context
+		DB     *database.DB
+		Logger *logrus.Logger
+		// If set, this cookie will be attached to all requests.
+		Cookie *http.Cookie
+
+		cancel context.CancelFunc
+	}
+)
+
+// ExtractCookie is a helper method which extracts the login cookie from a
+// response, so we can use it with future requests while testing.
+func ExtractCookie(r *http.Response) *http.Cookie {
+	for _, c := range r.Cookies() {
+		if c.Name == api.CookieName {
+			return c
+		}
+	}
+	return nil
+}
+
+// NewAccountsTester creates and starts a new AccountsTester service.
+// Use the Shutdown method for a graceful shutdown.
+func NewAccountsTester() (*AccountsTester, error) {
+	ctx, cancel := context.WithCancel(context.Background())
+	logger := logrus.New()
+
+	// Initialise the environment.
+	email.ServerLockID = "siasky.test"
+	email.PortalAddress = testPortalAddr
+	jwt.JWTPortalName = testPortalAddr
+	jwt.AccountsJWKSFile = pathToJWKSFile
+	err := jwt.LoadAccountsKeySet(logger)
+	if err != nil {
+		cancel()
+		return nil, errors.AddContext(err, fmt.Sprintf("failed to load JWKS file from %s", jwt.AccountsJWKSFile))
+	}
+
+	// Connect to the database.
+	db, err := database.New(ctx, DBTestCredentials(), logger)
+	if err != nil {
+		cancel()
+		return nil, errors.AddContext(err, "failed to connect to the DB")
+	}
+
+	// Start a noop mail sender in a background thread.
+	sender, err := email.NewSender(ctx, db, logger, &DependencySkipSendingEmails{}, FauxEmailURI)
+	if err != nil {
+		cancel()
+		return nil, errors.AddContext(err, "failed to create an email sender")
+	}
+	sender.Start()
+
+	// The meta fetcher will fetch metadata for all skylinks. This is needed, so
+	// we can determine their size.
+	mf := metafetcher.New(ctx, db, logger)
+
+	// The server API encapsulates all the modules together.
+	server, err := api.New(db, mf, logger, email.NewMailer(db))
+	if err != nil {
+		cancel()
+		return nil, errors.AddContext(err, "failed to build the API")
+	}
+
+	// Start the HTTP server in a goroutine and gracefully stop it once the
+	// cancel function is called and the context is closed.
+	srv := &http.Server{
+		Addr:    ":" + testPortalPort,
+		Handler: server.Router(),
+	}
+	go func() {
+		_ = srv.ListenAndServe()
+	}()
+	go func() {
+		select {
+		case <-ctx.Done():
+			_ = srv.Shutdown(context.TODO())
+		}
+	}()
+
+	return &AccountsTester{
+		Ctx:    ctx,
+		DB:     db,
+		Logger: logger,
+		cancel: cancel,
+	}, nil
+}
+
+// Get executes a GET request against the test service.
+func (at *AccountsTester) Get(endpoint string, params map[string]string) (r *http.Response, body []byte, err error) {
+	return at.executeRequest(http.MethodGet, endpoint, params, nil)
+}
+
+// Delete executes a DELETE request against the test service.
+func (at *AccountsTester) Delete(endpoint string, params map[string]string) (r *http.Response, body []byte, err error) {
```

Comment:
I'm not yet sure why you'd need to expose low-level endpoints like `Delete`, `Post` and `Put` instead of creating helpers which talk to a specific endpoint and are named accordingly but either way I'd just replace `params` with a `url.Values` if you are going to encode them anyway.

Reply:
Low-level access: I can have a small and really useful tester instead of dozens and dozens of repetitive helpers. I really prefer this approach.

`url.Values`: I chose `map[string]string` for its flexibility and the ability to quickly inline it within the method calls. But I don't mind using `url.Values`. A lot more typing for no real gain, IMO.